src/expr/src/relation/mod.rs

```diff
@@ -455,7 +455,71 @@ impl MirRelationExpr {
                             .unwrap();
                     }
                 }
-                RelationType::new(base_cols)
+
+                // Generally, unions do not have any unique keys, because
+                // each input might duplicate some. However, there is at
+                // least one idiomatic structure that does preserve keys,
+                // which results from SQL aggregations that must populate
+                // absent records with default values. In that pattern,
+                // the union of one GET with its negation, which has first
+                // been subjected to a projection and map, we can remove
+                // their influence on the key structure.
+                //
+                // If there are A, B, each with a unique `key` such that
+                // we are looking at
+                //
+                //     A + (B - A.proj(key)).map(stuff)
+                //
+                // Then we can report `key` as a unique key.
+                //
+                // TODO: make unique key structure an optimization analysis
+                // rather than part of the type information.
+                // TODO: perhaps ensure that (above) A.proj(key) is a
+                // subset of B, as otherwise there are negative records
+                // and who knows what is true (not expected, but again
+                // who knows what the query plan might look like).
```

Comment:
Is this why are we looking for such a complicated pattern instead of just `B-A`, where B has unique key <key> and A also has the same unique key <key>?

Reply:
I'm not sure I follow exactly, but .. if A is not a subset of B then there could be negative records.